### ISSUE TITLE
Question: high level program flow

### ISSUE THREAD

**Issue by duanjason** [skylot/jadx#388]
Hi Skylot,

Is there any high level documentation on how the program works?

So far, my understanding on the flow from reading various articles is:
apk --> DEX bytecode --> smali --> Java source code.

Is this correct? Does Java bytecode exist at any point during the process?

Thanks

**Comment by jpstotz** [comment]
What do you mean with "Does Java bytecode exist"? The DEX byte code is stored in the DEX files within the APK files. Therefore the DEX byte code exists of course, just unpack the APK and you get the DEX files with the DEX byte code.

**Comment by skylot** [comment]
@duanjason sorry but there is no any documentation for jadx (I'm so lazy and it is a lot of work).
Actually, jadx don't invent any new technics of decompilation so you can read any general decompiler architecture. For example: [Anatomy of a Java Decompiler](https://accu.org/index.php/journals/1850) (not original link of article mention [here](https://twitter.com/mstrobel/status/410394978477932544)).

By "Java bytecode" did you mean original java stack base bytecode? If yes, then all such code was converted to dex bytecode (using dx library) during building `classes.dex` and apk. Dex bytecode is register base and was designed for original [Dalvik VM](https://en.wikipedia.org/wiki/Dalvik_(software)). So original java bytecode doesn't exist in the decompilation process, but dex bytecode is very similar.

In short jadx decompilation flow contains several transformations and optimization passes ([code](https://github.com/skylot/jadx/blob/master/jadx-core/src/main/java/jadx/core/Jadx.java#L57)):
- read instructions list of every method
- split into basic blocks
- build control flow graph (CFG)
- run analysis steps (like SSA transform, loops detection, dominance frontier)
- a lot of optimizations on instructions to fill missing information and desugar code (type inference, code shrinker)
- place basic blocks into the recursive structure of code blocks ('regions' in jadx terms)
- some regions optimizations
- generate final code

**Comment by skylot** [comment]
1. No, jadx don't use original java bytecode, conversion is direct: from dex to the source code.
2. Compilation of generated code in most cases will **not** work but sometimes is possible. Jadx not perfect and a lot of errors occurs, so some methods not decompiled or just incorrect. All third party libraries exist in apk and no additional dependencies need. But sometimes to rebuild project is easier to use original libraries instead of decompiled (because of jadx errors).

**Comment by duanjason** [comment]
Thanks. I learnt that ["Dex2jar is dead, long live Jadx the Android decompiler"](https://www.facebook.com/riisllc/posts/dex2jar-is-dead-long-live-jadx-the-android-decompiler). I need to extract the .class files from apk. They don't need to be identical to the original .class files, but they need to be functional the same. Are you ware of anything that will do the job beside dex2jar?

**Comment by skylot** [comment]
@duanjason you can try [enjarify](https://github.com/Storyyeller/enjarify). Some time ago it was better than dex2jar, but now it looks outdated :(

**Comment by duanjason** [comment]
@skylot Thanks! Are you aware of any libraries that can used to modify classes in the classes.dex file? Something similar to ASM to modify Java class files.

**Comment by skylot** [comment]
@duanjason I know about dexlib2 from smali ([github](https://github.com/JesusFreke/smali/tree/master/dexlib2), [maven](https://mvnrepository.com/artifact/org.smali/dexlib2/2.2.5)) but it lack documentation. Hope it helps.

**Comment by duanjason** [comment]
@skylot Thanks!

**Comment by S-trace** [comment]
Maybe a bit offtopic, but sometimes I use dx to compile .jar file to .dex file, and then run jadx on that .dex file - jadx produces better result comparing to other java decompilers

**Comment by skylot** [comment]
@S-trace thanks for that :+1: 
And you can directly open `.jar` files in jadx and it will be converted using dx automatically.

P.S. Did you try [CFR](http://www.benf.org/other/cfr/)? It supports Java 8+ and works really good, I often compare it with jadx :)

**Comment by S-trace** [comment]
> And you can directly open .jar files in jadx and it will be converted using dx automatically.
Thank you, i didn't know about this feature!
> P.S. Did you try CFR? It supports Java 8+ and works really good, I often compare it with jadx :)
No, but now i will, thank you!)

**Comment by duanjason** [comment]
@S-trace Refer to your comment, "but sometimes I use dx to compile .jar file to .dex file", how you get the jar file? Is it produced by Android Studio?

**Comment by S-trace** [comment]
@duanjason 
> how you get the jar file? Is it produced by Android Studio?

No, it's regular Java jar file (for example, jadx.jar)
